### PR TITLE
SONAR-7961 Not possible to update quality profile of a project

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 1]
```diff
@@ -415,6 +376,7 @@ protected void configureLevel() {
       QProfileBackuper.class,
       QProfileReset.class,
       RubyQProfileActivityService.class,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: MINOR Remove this use of "RubyQProfileActivityService"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* CRITICAL 4 critical
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [QualityProfileDao.java#L180](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L180): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [QualityProfileDao.java#L220](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L220): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [QualityProfileDao.java#L305](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L305): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [SearchAction.java#L172](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchAction.java#L172): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. MAJOR [QualityProfileDao.java#L177](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L177): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [QualityProfileDao.java#L217](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L217): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [QualityProfileDao.java#L302](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L302): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/44950acc9b0db5b7a960601586970dfb4ae6ea45/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/ProjectsActionTest.java [hunk 2]
```diff
@@ -71,28 +63,18 @@
...
-  System2 system2 = mock(System2.class);
+  private WsTester wsTester = new WsTester(new QProfilesWs(
+    mock(RuleActivationActions.class),
+    mock(BulkRuleActivationActions.class),
+    mock(ProjectAssociationActions.class),
+    new ProjectsAction(dbClient, userSessionRule)));
 
   @Before
   public void setUp() {
-    dbTester.truncateTables();
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/ProjectsActionTest.java]: 👍

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 3]
```diff
@@ -214,28 +214,10 @@
...
 import org.sonar.server.qualityprofile.ws.OldRestoreAction;
 import org.sonar.server.qualityprofile.ws.ProfilesWs;
-import org.sonar.server.qualityprofile.ws.ProjectAssociationActions;
-import org.sonar.server.qualityprofile.ws.ProjectsAction;
-import org.sonar.server.qualityprofile.ws.QProfilesWs;
-import org.sonar.server.qualityprofile.ws.RenameAction;
-import org.sonar.server.qualityprofile.ws.RestoreAction;
-import org.sonar.server.qualityprofile.ws.RestoreBuiltInAction;
-import org.sonar.server.qualityprofile.ws.RuleActivationActions;
+import org.sonar.server.qualityprofile.ws.QProfilesWsModule;
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: 👍 !

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 4]
```diff
@@ -415,6 +376,7 @@ protected void configureLevel() {
       QProfileBackuper.class,
       QProfileReset.class,
       RubyQProfileActivityService.class,
+      QProfilesWsModule.class,
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: 👍 !

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import javax.annotation.Nullable;
+import org.sonar.server.component.ComponentService;
+import org.sonar.server.exceptions.NotFoundException;
+import org.sonar.server.qualityprofile.QProfile;
+import org.sonar.server.qualityprofile.QProfileLookup;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+
+public class ProjectAssociationFinder {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java]: I fear the merge with Simon's work will not be a pleasant task @simonbrandhof

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  private final ComponentService componentService;
+
+  public ProjectAssociationFinder(QProfileLookup profileLookup, ComponentService componentService) {
+    this.profileLookup = profileLookup;
+    this.componentService = componentService;
+  }
+
+  public String getProfileKey(@Nullable String language, @Nullable String profileName, @Nullable String profileKey) {
+    checkArgument((!isEmpty(language) && !isEmpty(profileName)) ^ !isEmpty(profileKey), BAD_PROFILE_PARAMETERS_ERROR);
+    if (profileKey == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java]: suggestion: inline

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      return getProfileKeyFromLanguageAndName(language, profileName);
+    }
+    return profileKey;
+  }
+
+  public String getProjectUuid(@Nullable String projectKey, @Nullable String projectUuid) {
+    checkArgument(!isEmpty(projectKey) ^ !isEmpty(projectUuid), BAD_PROJECT_PARAMETERS_ERROR);
+    if (projectUuid == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java]: suggestion: inline

File: sonar-ws/src/main/java/org/sonarqube/ws/client/qualityprofile/RemoveProjectRequest.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+public class RemoveProjectRequest {
+
+  private final String language;
+  private final String profileName;
+  private final String profileKey;
+  private final String projectKey;
+  private final String projectUuid;
+
+  public RemoveProjectRequest(Builder builder) {
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/qualityprofile/RemoveProjectRequest.java]: remark: now that we use this request internally only, we can add more intelligence. For example, we could imagine to have a QProfileRef which take a profileKey, a profileName and a language and returns a QProfileRef which guarantees the non nullability of its parameters. The QProfileRef objects has been done by Simon in a parellel PR

File: sonar-ws/src/test/java/org/sonarqube/ws/client/qualityprofile/QualityProfilesServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.sonarqube.ws.client.WsConnector;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_DEFAULTS;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_LANGUAGE;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_PROFILE_NAME;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_PROJECT_KEY;
+
+public class QualityProfilesServiceTest {
```
Comment (teryk) [inline sonar-ws/src/test/java/org/sonarqube/ws/client/qualityprofile/QualityProfilesServiceTest.java]: :+1

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationFinder.java]: Indeed, it was already there, I just move the code outside the class. Don't worry, I'll handle the merge, I've seen worst !

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java [hunk 10]
```diff
@@ -351,8 +351,8 @@ public void deleteProjectProfileAssociation(String projectUuid, String profileKe
     mapper(session).deleteProjectProfileAssociation(projectUuid, profileKey);
   }
 
-  public void updateProjectProfileAssociation(String projectUuid, String profileKey, DbSession session) {
-    mapper(session).updateProjectProfileAssociation(projectUuid, profileKey);
+  public void updateProjectProfileAssociation(String projectUuid, String newProfileKey, String oldProfileKey, DbSession session) {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: so now SQ really handles multi languages properly : )

File: sonar-ws/src/main/java/org/sonarqube/ws/client/qualityprofile/RemoveProjectRequest.java (hunk 8 above)
Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/qualityprofile/RemoveProjectRequest.java]: I'll try to use it in the merge

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java (hunk 10 above)
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: Yes, it's incredible that this issue was not reported sooner !

**Comment by teryk** [comment]
lgtm-no-comment-awesome

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java (hunk 10 above)
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: It tells us that we need to level up our game to help users update their quality profiles!

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* CRITICAL 4 critical
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [QualityProfileDao.java#L180](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L180): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [QualityProfileDao.java#L220](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L220): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [QualityProfileDao.java#L305](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L305): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. CRITICAL [SearchAction.java#L172](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchAction.java#L172): Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)
1. MAJOR [QualityProfileDao.java#L177](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L177): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [QualityProfileDao.java#L217](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L217): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [QualityProfileDao.java#L302](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L302): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/a56e60d168802ff9dc23ffec978bb412fd501f97/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

**Pull request merged**